### PR TITLE
VW PQ: Volkswagen Passat NMS

### PR DESCRIPTION
**Pull request by jyoung8607**
- [x] commaai/opendbc#630
- [x] Clean up LDW_1 message in opendbc
- [x] Run checksums in opendbc CC packer instead of openpilot
- [x] commaai/opendbc#633
- [x] commaai/panda#964
- [x] Fix stock ACC cancel spamming
- [x] Add support for HUD message (driver takeover request)
- [x] Read vEgoRaw from alternate source, to support NSF later (and match current Panda)
- [x] Read non-driver door and trunk states, if possible
- [x] ~~Read stock FCW/AEB states~~
- [x] Read stock LDW states and forward to SWA for camera integrations
- [x] Test route for Panda safety
- [x] Test route for openpilot
- [x] Set dashcamOnly prior to release from draft

Some of the abstraction and cleanup work (particularly CC button handling) may get done in the MQB port first.

### CODE REVIEW

File: selfdrive/car/volkswagen/pqcan.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+def create_pq_steering_control(packer, bus, apply_steer, idx, lkas_enabled):
+  values = {
+    "HCA_Zaehler": idx,
+    "LM_Offset": abs(apply_steer),
+    "LM_OffSign": 1 if apply_steer < 0 else 0,
+    "HCA_Status": 5 if (lkas_enabled and apply_steer != 0) else 3,
```
Comment (GregorKikelj) [inline selfdrive/car/volkswagen/pqcan.py:6]: Hca 7 with torque controller would be cool to have in the future

**Comment by actuallylemoncurd** [comment]
>     * [ ]  Read non-driver door and trunk states

`
BO_ 1136 BSG_Kombi: 5 XXX
 SG_ Heckdeckel_geoeffnet : 13|1@1+ (1,0) [0|0] "" XXX
 SG_ Motorhaube_geoeffnet : 12|1@1+ (1,0) [0|0] "" XXX
 SG_ Tuer_hinten_rechts_geoeffnet : 11|1@1+ (1,0) [0|0] "" XXX
 SG_ Tuer_hinten_links_geoeffnet : 10|1@1+ (1,0) [0|0] "" XXX
 SG_ Beifahrertuer_geoeffnet : 9|1@1+ (1,0) [0|0] "" XXX
 SG_ Fahrertuer_geoeffnet : 8|1@1+ (1,0) [0|0] "" XXX
`

This should work, or atleast I think so. Can verify once C3 arrives.

**Comment by jyoung8607** [comment]
> This should work, or atleast I think so. Can verify once C3 arrives. Bonus! Theyre already in the DBC! Just need to tack them into carstate.

It's in the DBC, but we need to make sure it shows up on the CAN buses where we need it. It might show up on extended CAN, but not older vehicles where we'd integrate with a powertrain tap. Or, it might only show up on the PQ dedicated cluster CAN or convenience CAN, and not be available to us at all.

A copy of the driver's door state gets repeated as part of an aggregate message by the gateway for the exact reason you mentioned, repriming the fuel system. I'm not sure if the rest will be available on powertrain, or even extended. I would very much like to write this to be compatible with future powertrain-only integrations, but it will depend on tradeoffs like this.

**Comment by jyoung8607** [comment]
Per my discussion with @adeebshihadeh, I'm retargeting this PR to a minimum drivable port that can be considered for merge in dashcam mode and moving some of the original tasks to future PRs.

To exit from dashcam mode:
- [ ] KWP2000 over VW CAN-TP 2.0 support for EPS firmware identification and other queries
- [ ] Find a way to query the EPS to see if the six minute timeout/steer to zero workarounds are applied
- [ ] Bring Panda safety out from under ALLOW_DEBUG

Don't actually care about this until VW PQ openpilot longitudinal, but don't want to forget:
- [ ] Figure out how short-press and long-press speed increments will work with PQ's separate signals

**Comment by actuallylemoncurd** [comment]
> I'm retargeting this PR to a minimum drivable port that can be considered for merge in dashcam mode

I see you have some todo's listed to exit from dashcam mode, what are the todo's to get dashcam upstreamed into comma?

**Comment by jyoung8607** [comment]
> I see you have some todo's listed to exit from dashcam mode, what are the todo's to get dashcam upstreamed into comma?

My own list is what remains in the opening PR comment. TL;DR: fix cancel spam, do some other finishing and polishing, drive it and get test routes. After that, comma may have their own thoughts.

**Comment by CineXWorm** [comment]
> > This should work, or atleast I think so. Can verify once C3 arrives. Bonus! Theyre already in the DBC! Just need to tack them into carstate.
> 
> It's in the DBC, but we need to make sure it shows up on the CAN buses where we need it. It might show up on extended CAN, but not older vehicles where we'd integrate with a powertrain tap. Or, it might only show up on the PQ dedicated cluster CAN or convenience CAN, and not be available to us at all.
> 
> A copy of the driver's door state gets repeated as part of an aggregate message by the gateway for the exact reason you mentioned, repriming the fuel system. I'm not sure if the rest will be available on powertrain, or even extended. I would very much like to write this to be compatible with future powertrain-only integrations, but it will depend on tradeoffs like this.

These are present on my extended canbus.

**Comment by jyoung8607** [comment]
> These are present on my extended canbus. The Gate_Komf_1 message is actually being combined from other messages aswell by the gateway. Same as mACC_System on non ACC cars.

You see `BSG_Kombi` on Extended CAN? Do you have an uploaded rlog you can show us? The trouble is `Gate_Komf_1` only has the driver's door, or at least that's the only documented door it has.

That said, even if it's there for you, @actuallylemoncurd says it isn't present on Extended CAN on his Mk6 JSW. If it's not something we can depend on globally, it's not of much use.

**Comment by jyoung8607** [comment]
Updates:

* Have found passenger door, rear door, and trunk signals that should work on most cars we care about
* LDW_1 updates and message forwarding are done
* Checksum and counter work is done and ready for consideration
* Deferring stock FCW/AEB signal handling until I understand how this works on pre-MY2016 cars

Remaining:
* CC button spam cleanup, will tie this in with MQB cleanup per my DMs with @sshane
* Drive testing, I'm told this is engaging but not steering for some test users, will investigate

**Comment by jyoung8607** [comment]
This week is the first time I've personally driven this code/vehicle.

Bad news, it turns out to have ACC "basic" with a nonzero minimum accel speed, which I thought was only a problem with older vehicles. Also, the PIF tune was a little janky. Good news, I stubbed in a guess for new-hotness torque control parameters and the car responded *very* positively. It definitely has more available lateral accel than the MQB vehicles I've driven, and I haven't even tried the legendary "HCA 7" mode yet.

CC button spam worked right away with code based on the #25168 cleanup, and the port is otherwise driving okay after a couple other minor bugfixes. I'm not ready for review yet, I'm not fully happy with the code and abstractions, but I'm ready for test routes for CI. We already have one for Panda.

@adeebshihadeh can you make `3cfdec54aa035f3f|2022-07-19--23-45-10` available in the openpilot CI test bucket(s)?

**Comment by jyoung8607** [comment]
Ready for review. Notes and known issues:

* Dashcam-only at this time due to the famous six-minute timebomb. Removing from dashcam is contingent on further work to enable KWP2000 fingerprinting of the EPS, so we can auto-identify cars with fixed calibrations.
 * There doesn't appear to be a silent HUD "take control" alert. The supposedly silent one still beeps on at least some tested vehicles. More testing needs to be done. We may remove the HUD message if no alternative can be found.
* Docs footnote for NMS-vs-B8 Passat differentiation has been removed because the docs generator doesn't currently know how to exclude footnotes that only apply to dashcamOnly cars, and it would take a minor refactor to fix.
* Longitudinal control is proof-of-concept only. It is VOACC (no radar points) and only very basic tuning has been done, but it's functional for highway driving. There are known issues with HUD messaging (FCW alerts, HUD focus being stuck on ACC), which should be fixable. Follow-to-stop isn't supported yet because the development car didn't support it.
* Stock FCW/AEB detection in carstate should be re-added once long control false FCW/AEB issues are fixed.

File: selfdrive/car/volkswagen/interface.py [hunk 2]
```diff
@@ -25,7 +28,36 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), car_fw=None, disa
...
+
+      # The PQ port is in dashcam-only mode due to a fixed six-minute maximum timer on HCA steering. An unsupported
+      # EPS flash update to work around this timer, and enable steering down to zero, is available from:
+      #   https://github.com/pd0wm/pq-flasher
+      # It is documented in a four-part blog series:
+      #   https://blog.willemmelching.nl/carhacking/2022/01/02/vw-part1/
+      # Panda ALLOW_DEBUG firmware required.
+      ret.dashcamOnly = True
+
+      if Params().get_bool("DisableRadar") and ret.networkLocation == NetworkLocation.gateway:
```
Comment (adeebshihadeh) [inline selfdrive/car/volkswagen/interface.py:54]: the param is passed in
```suggestion
      if disable_radar and ret.networkLocation == NetworkLocation.gateway:
```

File: selfdrive/car/volkswagen/carstate.py [hunk 3]
```diff
@@ -135,8 +138,111 @@ def update(self, pt_cp, cam_cp, ext_cp, trans_type):
...
+    ret.steeringAngleDeg = pt_cp.vl["Lenkhilfe_3"]["LH3_BLW"] * (1, -1)[int(pt_cp.vl["Lenkhilfe_3"]["LH3_BLWSign"])]
+    ret.steeringRateDeg = pt_cp.vl["Lenkwinkel_1"]["Lenkradwinkel_Geschwindigkeit"] * (1, -1)[int(pt_cp.vl["Lenkwinkel_1"]["Lenkradwinkel_Geschwindigkeit_S"])]
+    ret.steeringTorque = pt_cp.vl["Lenkhilfe_3"]["LH3_LM"] * (1, -1)[int(pt_cp.vl["Lenkhilfe_3"]["LH3_LMSign"])]
+    ret.steeringPressed = abs(ret.steeringTorque) > self.CCP.STEER_DRIVER_ALLOWANCE
+    ret.yawRate = pt_cp.vl["Bremse_5"]["Giergeschwindigkeit"] * (1, -1)[int(pt_cp.vl["Bremse_5"]["Vorzeichen_der_Giergeschwindigk"])] * CV.DEG_TO_RAD
+
+    # Verify EPS readiness to accept steering commands
+    hca_status = self.CCP.hca_status_values.get(pt_cp.vl["Lenkhilfe_2"]["LH2_Sta_HCA"])
+    ret.steerFaultPermanent = hca_status in ["DISABLED", "FAULT"]
+    ret.steerFaultTemporary = hca_status in ["INITIALIZING", "REJECTED"]
```
Comment (adeebshihadeh) [inline selfdrive/car/volkswagen/carstate.py:167]:
```suggestion
    ret.steerFaultPermanent = hca_status in ("DISABLED", "FAULT")
    ret.steerFaultTemporary = hca_status in ("INITIALIZING", "REJECTED")
```

File: selfdrive/car/volkswagen/carstate.py [hunk 4]
```diff
@@ -135,8 +138,111 @@ def update(self, pt_cp, cam_cp, ext_cp, trans_type):
...
+
+    # Update gas, brakes, and gearshift.
+    ret.gas = pt_cp.vl["Motor_3"]["Fahrpedal_Rohsignal"] / 100.0
+    ret.gasPressed = ret.gas > 0
+    ret.brake = pt_cp.vl["Bremse_5"]["Bremsdruck"] / 250.0  # FIXME: this is pressure in Bar, not sure what OP expects
+    ret.brakePressed = bool(pt_cp.vl["Motor_2"]["Bremstestschalter"])
+    self.parkingBrakeSet = bool(pt_cp.vl["Kombi_1"]["Bremsinfo"])
```
Comment (adeebshihadeh) [inline selfdrive/car/volkswagen/carstate.py:174]: unused?

Comment (jyoung8607) [reply selfdrive/car/volkswagen/carstate.py]: Good catch. It worked in the past, but that line wasn't updated to follow the refactor adding parkingBrake to CarState with common event processing. Fixed.

**Pull request merged**